gp_emulator/gpu/gpu_predict.h

```diff
@@ -0,0 +1,94 @@
+/* A file to test imorting C modules for handling arrays to Python */
+
+#include "Python.h"
+#include "arrayobject.h"
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+/* Includes, cuda */
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <cublas_v2.h>
+#define DOUBLE__PRECISION
+#define CUDA_BLOCK 2
+ 
+#ifdef DOUBLE__PRECISION
+  #define real double
+#else
+  #define real float
+#endif 
+
+#define IDX2D(i,j,ld) (((j)*(ld))+(i))//column major
+
+
+#ifdef DOUBLE__PRECISION
+  #define CUBLAS_GEMV cublasDgemv
+  #define CUBLAS_GEMM cublasDgemm
+  #define CUBLAS_GEAM cublasDgeam
+#else 
+  #define CUBLAS_GEMV cublasSgemv
+  #define CUBLAS_GEMM cublasSgemm
+  #define CUBLAS_GEAM cublasSgeam
+#endif
+
+
+
+PyArrayObject *pyvector(PyObject *objin);
+real*pyvector_to_Carrayptrs(PyArrayObject *arrayin);
+real **pymatrix_to_Carrayptrs(PyArrayObject *arrayin);
+real **ptrvector(long n);
+PyObject *predict_wrap ( PyObject *self, PyObject *args );
+
+
+void getPredictDataFromPython(PyObject *args, real **c_theta_exp, real **c_invQt, real **c_invQ, 
+                              real **c_testing, real **c_inputs,
+                              real **c_mu, real **c_var, real **c_deriv,
+                              int *N, int *M, int *D, int *theta_size);
+
+#ifdef __cplusplus
+extern "C"{
+#endif
+void predict(const real *c_theta_exp,const  real *c_inputs,const real *c_invQt, const real *c_invQ, const real *c_testing, real *c_mu, real *c_var, real *c_deriv, const int N, const int M, const int  D, const int theta_size);
+real *computeTranspose(const real *matrix, const  int size_in, const  int size_out);
+void gpu_vectorTimesMatrix(const real *A, const real *v, real *res, int nrows, int ncols);
+void gpu_init_array(real *vec, const int init_val, const int vec_len);
+void gpu_cdist(const real *input1, const real *input2, real *output, const int nrow1, const int ncol1, const int nrow2, const int ncol2);
+void gpu_matrixExp( real *matrix,const real alpha,const real beta, const int size );
+void gpu_elementwiseMult( const real *v1, real *v2, const int size );
+void gpu_scalarMinusVec( real *matrix, const real scalar, const int size );
+real* gpu_rowSum(const real *A, const int A_nrows,const int A_ncols);
+void gpu_crossMinus(const real *v1, const real *v2, real *mat_res, const int v1_len, const int v2_len);
+#ifdef __cplusplus
+}
+#endif
+
+
+
+// error check macros
```

Comment:
Put these in a separate header file. E.g. CUDA Macros.
Always use UPPER_CASE for Macros so we don't confuse them with functions.